### ISSUE TITLE
How to forward both stdin and stderr?

### ISSUE THREAD

**Issue by piegamesde** [lmorg/murex#389]
I'd like to run a command silently, i.e. pipe both stdin and stderr into `null`. But there are also other use cases where this is necessary.

How can I achieve this?

**Comment by lmorg** [comment]
Good question, this is a big area where the documentation is sorely lacking.

## Pipe tokens

There's a few ways to do redirections depending on your use case. If you wanted to swap STDOUT and STDERR around then you can pipe using the `?` token instead of `|` / `->` (https://murex.rocks/docs/parser/pipe-err.html). This doesn't solve your immediate problem though.

## Named pipes

You can also embed redirection rules into the command using angle brackets, eg
```
command <stdout> <!stderr> parameters...
```

In practice this would look like the following (though first I'm going to create an example function to illustrate things)
```
» function outerr { out @PARAMS; err "Error: message" }                                                                                                                                  
» outerr hello world                                                                                                                                                            
hello world
Error: message
```

So to use angle brackets for redirection you'd do the following
```
» outerr <null> <!null> hello world 
```

The item in the angle brackets:
- without a bang prefix is STDOUT (eg `<null>`)
- with a bang prefix is STDERR (eg `<!null>`)

These are called named pipes. `null` is a built in one so you can use with without defining it, but you can create your own named pipes too using the `pipe` command https://murex.rocks/docs/commands/pipe.html (the documentation on this is very poor because the CMS seems to have pulled most of the variables docs too. doh!)

Named pipes can be regular flows of data between processes but you can also use them for things like networking, eg
```perl
# Create the named pipe. In this instance it is an outbound TCP connection to Google on
# port 80. So basically a plain text HTTP connection.
pipe example --tcp-dial google.com:80

# Lets create a background process that outputs any data received from that named pipe.
# Reading from pipes are a blocking process. ie `out: "Pipe closed"` will not execute while
# the named pipe is open. In that regard it is a little like running `cat`.
bg {
    <example>
    out: "Pipe closed"
}

# Write a basic HTTP header "GET" request
out: "GET /" -> <example>

# At this stage we will see a HTML response due to the background (`bg`) read above.
# It doesn't matter that the HTML response is a redirect page, that's just because we're
# sending data in clear text for this example so Google is rightfully insisting we use TLS
# encryption for our traffic. But we don't need TLS for this hypothetical example.

# It is good practice to close named pipes when you're done with them to prevent any
# locking issues when writing mutli-tasking scripts.
!pipe: example
```

In that example above, you could also have `<example>` be a redirection. Eg if your HTTP message was a text file (eg `http.txt` you could have done something like:
```
cat <example> http.txt
```
to redirect the output of the text file to Google.

This is documented here: https://murex.rocks/docs/commands/namedpipe.html though again there it's not the clearest of documents.

This is also a way to dynamically create a named pipe and auto-close it again when that command exits. But I don't think I've gotten as far as documenting that at all yet. So I'll save that for another day.

## Additional uses of angle brackets

### Environmental variables

Additionally to named pipes, you can use them to set process specific environmental variables. eg
```
» env <env:foo=bar> | grep foo
foo=bar
```
(this is equivalent to `foo=bar env | grep foo` in Bash)
This hasn't yet been documented.

### Testing framework

As well as adding unit tests, watches and debugging messages in your code. This is also poorly documented: https://murex.rocks/docs/commands/test.html

# Solution

So to go back to your specific problem, you can specify the `null` named pipe like as follows
```
out: <null> <!null> This is just an example
```

**Comment by piegamesde** [comment]
Yeah this is probably a documentation issue. When I don't know how to do something in Bash, I can simply search it and there will very likely be someone on StackOverflow who had this problem before. For example, I regularly come back to this excellent answer regarding redirects: https://askubuntu.com/questions/420981/how-do-i-save-terminal-output-to-a-file/731237#731237

I think Murex should not only document the features themselves, but also how to use them; the idioms that solve common problems. Also, how do these named pipes handle failing commands? My original motivation was to test whether a command succeeds or not and then do things conditionally, but without having the output cluttered with the command. If the redirects change the behavior the same way some pipes do, this would not help my use case.

**Comment by lmorg** [comment]
> I think Murex should not only document the features themselves, but also how to use them; the idioms that solve common problems

Good suggestion. I'll take this on board.

> Also, how do these named pipes handle failing commands?

They don't handle anything. They're just a data pipe. A command will send data down it and if it produces any output then that output will be sent to whatever pipe is assigned to that output stream; but there's nothing about pipes that handle process execution. This is the same for Bash too. The advantage of pipes being a separate abstraction to process execution is that you can then handle process management independently from chaining commands together in pipelines.

> My original motivation was to test whether a command succeeds or not and then do things conditionally, but without having the output cluttered with the command. If the redirects change the behavior the same way some pipes do, this would not help my use case.

Neither redirects nor pipes (which are just different syntax for the same abstractions) should affect the behavior of a command, unless said command specifically queries the file descriptor it's writing any output too.

In 99% cases most programs will behave the same regardless of how it's output is piped because that's how CLI software was designed (STDOUT, STDERR etc are just file descriptors that programs write to. They shouldn't care what program then reads from those FDs). However some tools like to show different output when used in a terminal (TTY) vs piped because they'll output human readable content in the former and machine readable content in the latter. An example of this is `ls`
```
# compare the output of
ls

# vs the output of
ls | cat
```
You'll see the former is multi-column grid where as the latter is a single column list. Also this behavior will be the same regardless of whatever shell you use because it's behavior written into the code of `ls` itself. What's happening here is `ls` is testing whether the FD is a TTY or a pipe.

If you're testing the behavior of a command that works like this then you can always pipe the output into `cat` to see what it looks like when STDOUT is a TTY. Or maybe check if it has flags to define other machine readable output modes (eg `-o json` when using Kubernetes (`kubectl`), or `--porcelain` mode for Git's `git` CLI). The test library built into murex might help with your problem here too -- albeit that ostensibly just works as a wrapper around named pipes so would be subject to the same limitations if the developers of any command decided to write code that checks for whether the output is a TTY.

If you're able to share (ie there's no company secret source) the problem you're having in more detail I'm happy to help.

**Comment by piegamesde** [comment]
My use case is just testing whether I can `ping` some address or not. I was asking about exit behavior because I think when I tried `ping ? null` it looked like it swallowed the exit code.

**Comment by lmorg** [comment]
Yeah null would affect the exit code there because you’re invoking it as a command. You’re literally piping ping’s STDERR to a new command called null and thus the exit code you’re checking is nulls and not pings.

https://murex.rocks/docs/commands/devnull.html

The reason this command exists is purely legacy though. It was implemented long before this shell even hit GitHub. I needed a cross platform /dev/null device but hadn’t implemented names pipes yet.

Pragmatically speaking this command should probably be deprecated now however I was always nervous about removing it just in case it broke other peoples shell scripts. (Though I could probably replicate it as an alias to <null> and remove it from the docs so old scripts don’t break but new users don’t adopt it).
